pkg/controllers/custompackage/controller.go

```diff
@@ -77,55 +78,101 @@ func (r *Reconciler) reconcileCustomPackage(ctx context.Context, resource *v1alp
-	}
+	switch resource.Spec.ArgoCD.Type {
+	case argocdapplication.ApplicationKind:
+		app, ok := objs[0].(*argov1alpha1.Application)
+		if !ok {
+			return ctrl.Result{}, fmt.Errorf("object is not an ArgoCD application %s", resource.Spec.ArgoCD.ApplicationFile)
+		}
+
+		res, err := r.reconcileArgoCDApp(ctx, resource, app)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
 
-	appName := app.GetName()
-	if resource.Spec.Replicate {
-		synced := true
-		repoRefs := make([]v1alpha1.ObjectRef, 0, 1)
-		if app.Spec.HasMultipleSources() {
-			for j := range app.Spec.Sources {
-				s := &app.Spec.Sources[j]
-				res, repo, sErr := r.reconcileArgoCDSource(ctx, resource, s, appName)
-				if sErr != nil {
-					return res, sErr
+		foundAppObj := argov1alpha1.Application{}
+		err = r.Client.Get(ctx, client.ObjectKeyFromObject(app), &foundAppObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				err = r.Client.Create(ctx, app)
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("creating %s app CR: %w", app.Name, err)
 				}
-				if repo != nil {
-					if synced {
-						synced = repo.Status.InternalGitRepositoryUrl != ""
-					}
-					s.RepoURL = repo.Status.InternalGitRepositoryUrl
-					repoRefs = append(repoRefs, v1alpha1.ObjectRef{
-						Namespace: repo.Namespace,
-						Name:      repo.Name,
-						UID:       string(repo.ObjectMeta.UID),
-					})
+
+				return ctrl.Result{RequeueAfter: requeueTime}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("getting argocd application object: %w", err)
+		}
+
+		foundAppObj.Spec = app.Spec
+		foundAppObj.ObjectMeta.Annotations = app.GetAnnotations()
+		foundAppObj.ObjectMeta.Labels = app.GetLabels()
+		err = r.Client.Update(ctx, &foundAppObj)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("updating argocd application object %s: %w", app.Name, err)
+		}
+		return res, nil
+
+	case argocdapplication.ApplicationSetKind:
+		// application set embeds application spec. extract it then handle git generator repoURLs.
+		appSet, ok := objs[0].(*argov1alpha1.ApplicationSet)
+		if !ok {
+			return ctrl.Result{}, fmt.Errorf("object is not an ArgoCD application set %s", resource.Spec.ArgoCD.ApplicationFile)
+		}
+		res, err := r.reconcileArgoCDAppSet(ctx, resource, appSet)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		foundAppSetObj := argov1alpha1.ApplicationSet{}
+		err = r.Client.Get(ctx, client.ObjectKeyFromObject(appSet), &foundAppSetObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				err = r.Client.Create(ctx, appSet)
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("creating %s argocd application set CR: %w", appSet.Name, err)
 				}
+				return ctrl.Result{RequeueAfter: requeueTime}, nil
 			}
-		} else {
-			s := app.Spec.Source
-			res, repo, sErr := r.reconcileArgoCDSource(ctx, resource, s, appName)
+			return ctrl.Result{}, fmt.Errorf("getting argocd application set object: %w", err)
+		}
+
+		foundAppSetObj.Spec = appSet.Spec
```

Comment:
the reconcilliation logic is duplicated here with the above case. Is it possible to dedupe?

Reply:
I considered it but I don't think we'd save any lines of code or performance because we'd have to type cast in the new function then branch out anyways. I figured a bit of duplication here is acceptable.